### PR TITLE
Implement Zicond extension

### PR DESCRIPTION
**Pull request by rewired-gh**
**Related issue**: 
Not an issue.

**Type of change**: feature request

**Impact**: API addition (no impact on existing code)

**Development Phase**: implementation

**Release Notes**

Only implemented Zicond extension for `ALU` module. Still need to implement Zicond extension for other modules such as `ABLU` in the future.

By default, this extension is disabled. (Please refer to `useConditionalZero` in `RocketCoreParams`)

Also, `Instructions.scala` was updated. Code generated by `risv-opcodes` with command `EXTENSIONS="unratified/rv_zicond" make inst.chisel`. Only new lines were added.

### CODE REVIEW

File: src/main/scala/tile/Core.scala [hunk 1]
```diff
@@ -29,6 +29,7 @@ trait CoreParams {
   val useCryptoNIST: Boolean
   val useCryptoSM: Boolean
   val useRVE: Boolean
+  val useConditionalZero: Boolean
```
Comment (ZenithalHourlyRate) [inline src/main/scala/tile/Core.scala:32]: You should also add a `usingConditionalZero` in HasCoreParameters, then you can substitute all the `coreParams.useConditionalZero` to `usingConditionalZero`.

As of now, Zicond is not implemented in ABLU, so you may also add a requirement in `HasRocketCoreParameters` that `require(!(usingABLU && usingConditionalZero), "Zicond is not implemented in ABLU")` and remove such requirement when implementing that for ABLU.

File: src/main/scala/rocket/IDecode.scala [hunk 2]
```diff
@@ -111,6 +111,8 @@ class IDecode(aluFn: ALUFN = ALUFN())(implicit val p: Parameters) extends Decode
     SLL->       List(Y,N,N,N,N,N,Y,Y,N,N,N,N,A2_RS2, A1_RS1, IMM_X, DW_XPR,aluFn.FN_SL,    N,M_X,        N,N,N,N,N,N,Y,CSR.N,N,N,N,N),
     SRL->       List(Y,N,N,N,N,N,Y,Y,N,N,N,N,A2_RS2, A1_RS1, IMM_X, DW_XPR,aluFn.FN_SR,    N,M_X,        N,N,N,N,N,N,Y,CSR.N,N,N,N,N),
     SRA->       List(Y,N,N,N,N,N,Y,Y,N,N,N,N,A2_RS2, A1_RS1, IMM_X, DW_XPR,aluFn.FN_SRA,   N,M_X,        N,N,N,N,N,N,Y,CSR.N,N,N,N,N),
+    CZERO_EQZ-> List(Y,N,N,N,N,N,Y,Y,N,N,N,N,A2_RS2, A1_RS1, IMM_X, DW_XPR,aluFn.FN_CZEQZ, N,M_X,        N,N,N,N,N,N,Y,CSR.N,N,N,N,N),
+    CZERO_NEZ-> List(Y,N,N,N,N,N,Y,Y,N,N,N,N,A2_RS2, A1_RS1, IMM_X, DW_XPR,aluFn.FN_CZNEZ, N,M_X,        N,N,N,N,N,N,Y,CSR.N,N,N,N,N),
```
Comment (ZenithalHourlyRate) [inline src/main/scala/rocket/IDecode.scala:115]: These two lines should be separated into another class `IConditionalZeroDecode` and should be enabled by `usingConditionalZero` in the `decode_table` of RocketCore.scala

File: src/main/scala/rocket/ALU.scala [hunk 3]
```diff
@@ -160,10 +162,19 @@ class ALU(implicit p: Parameters) extends AbstractALU(new ALUFN)(p) {
   val shout = Mux(io.fn === aluFn.FN_SR || io.fn === aluFn.FN_SRA, shout_r, 0.U) |
               Mux(io.fn === aluFn.FN_SL,                           shout_l, 0.U)
 
+  // CZEQZ, CZNEZ
+  val cond_out = Option.when(coreParams.useConditionalZero)(
+    Mux(io.fn === aluFn.FN_CZEQZ && io.in2 =/= 0.U, io.in1, 0.U) |
+    Mux(io.fn === aluFn.FN_CZNEZ && io.in2 === 0.U, io.in1, 0.U)
```
Comment (ZenithalHourlyRate) [inline src/main/scala/rocket/ALU.scala:168]: Too many MUXes and compare. 

```scala
val in2_not_zero = io.in2.orR
Mux(io.fn === aluFn.FN_CZEQZ && in2_not_zero || io.fn === aluFn.FN_CZNEZ && !in2_not_zero, io.in1, 0.U)
```

This should be enough

Comment (ZenithalHourlyRate) [reply src/main/scala/rocket/ALU.scala:168]: @jerryz123 Are you happy with directly modifying ALU or should we add another ALU like ALUZicond? The main problem with the latter way is that then we will have a bunch of ALU  to maintain like `ALU`, `ALUZicond`, `ABLU`, `ABLUZicond`.

**Review by ZenithalHourlyRate — CHANGES_REQUESTED** [review]

**Comment by ZenithalHourlyRate** [comment]
You might try to test your code using riscv-arch-test as the reference model (https://github.com/riscv-software-src/riscv-isa-sim/pull/1241) and tests (https://github.com/riscv-non-isa/riscv-arch-test/pull/321) are available.

File: src/main/scala/rocket/ALU.scala (hunk 3 above)
Comment (jerryz123) [reply src/main/scala/rocket/ALU.scala:168]: This approach is fine.

**Comment by rewired-gh** [comment]
I'm sorry for the force push. It was because the email of the previous commit was mistyped and could not pass EasyCLA.

File: src/main/scala/rocket/ALU.scala [hunk 4]
```diff
@@ -160,10 +162,19 @@ class ALU(implicit p: Parameters) extends AbstractALU(new ALUFN)(p) {
...
+  )
+
   // AND, OR, XOR
   val logic = Mux(io.fn === aluFn.FN_XOR || io.fn === aluFn.FN_OR, in1_xor_in2, 0.U) |
               Mux(io.fn === aluFn.FN_OR || io.fn === aluFn.FN_AND, io.in1 & io.in2, 0.U)
-  val shift_logic = (aluFn.isCmp(io.fn) && slt) | logic | shout
+  val shift_logic = cond_out match {
+    case Some(co) => (aluFn.isCmp (io.fn) && slt) | logic | shout | co
+    case _ => (aluFn.isCmp (io.fn) && slt) | logic | shout
+  }
```
Comment (ZenithalHourlyRate) [inline src/main/scala/rocket/ALU.scala:177]: These two lines are somewhat duplicate.

```
  val shift_logic = (aluFn.isCmp(io.fn) && slt) | logic | shout
  val shift_logic_cond = cond_out match {
    case Some(co) => shift_logic | co
    case _ => shift_logic
  }
```

I prefer this way

**Review by ZenithalHourlyRate — COMMENTED** [review]
Other parts are fine to me now.

From private communication, it seems that it is not mature enough to add arch-test CI on Zicond as that test PR has not been merged and the toolchain does not have the support for it.

Adding CI for Zicond might be put in another draft PR.

**Comment by rewired-gh** [comment]
It seems that this implementation might have some problems.

As mentioned here:

> Other parts are fine to me now.
> 
> From private communication, it seems that it is not mature enough to add arch-test CI on Zicond as that test PR has not been merged and the toolchain does not have the support for it.
> 
> Adding CI for Zicond might be put in another draft PR.

So I've tried to debug with RBB emulator. I followed the instructions in `README.md` and write a simple assembly program, which contains the following instructions:

```assembly
addi x2, x0, 12    # Non-zero
add x3, x0, x0     # Zero
.word 0b00001110001000010101001000110011 # czero.eqz x4, x2, x2 (x4 := x2)
.word 0b00001110001100010101001000110011 # czero.eqz x4, x2, x3 (x4 := 0)
.word 0b00001110001000010111001000110011 # czero.nez x4, x2, x2 (x4 := 0)
.word 0b00001110001100010111001000110011 # czero.nez x4, x2, x3 (x4 := x2)
```

The first two instructions can be executed correctly. However, it seems that there are some problems when executing the third instruction. Here are related commands and output in GDB:

```bash
(gdb) stepi
0x0000000080000006      13          add x3, x0, x0     # Zero
   0x0000000080000002 <entry+2>:        000001b3                add     gp,zero,zero
=> 0x0000000080000006 <entry+6>:        0e215233                .insn   4, 0x0e215233
   0x000000008000000a <entry+10>:       0e315233                .insn   4, 0x0e315233
   0x000000008000000e <entry+14>:       0e217233                .insn   4, 0x0e217233
   0x0000000080000012 <entry+18>:       0e317233                .insn   4, 0x0e317233
(gdb) info registers x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0x9e55add3c2d5d9fc  0x9e55add3c2d5d9fc
(gdb) stepi
0x0000000000000000 in ?? ()
=> 0x0000000000000000:  0000                    unimp
1: x/6i $pc
=> 0x0: unimp
   0x2: unimp
   0x4: unimp
   0x6: unimp
   0x8: unimp
   0xa: unimp
(gdb) info registers x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0x9e55add3c2d5d9fc  0x9e55add3c2d5d9fc
```

**Comment by rewired-gh** [comment]
As mentioned in

> Other parts are fine to me now.
> 
> From private communication, it seems that it is not mature enough to add arch-test CI on Zicond as that test PR has not been merged and the toolchain does not have the support for it.
> 
> Adding CI for Zicond might be put in another draft PR.

So I've follow the instructions in README.md to debug the implementation with GDB. Here is the assembly code I used:

```assembly
addi x2, x0, 12    # Non-zero
add x3, x0, x0     # Zero
.word 0b00001110001000010101001000110011 # czero.eqz x4, x2, x2 (x4 := x2)
.word 0b00001110001100010101001000110011 # czero.eqz x4, x2, x3 (x4 := 0)
.word 0b00001110001000010111001000110011 # czero.nez x4, x2, x2 (x4 := 0)
.word 0b00001110001100010111001000110011 # czero.nez x4, x2, x3 (x4 := x2)
```

And here is the related GDB commands and output:

```bash
(gdb) i r x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0xffffffffb5e85757  0xffffffffb5e85757
(gdb) stepi       
0x000000008000000a      13          add x3, x0, x0     # Zero
   0x0000000080000002 <entry+2>:        000001b3                add     gp,zero,zero
   0x0000000080000006 <entry+6>:        0e215233                .insn   4, 0x0e215233
=> 0x000000008000000a <entry+10>:       0e315233                .insn   4, 0x0e315233
   0x000000008000000e <entry+14>:       0e217233                .insn   4, 0x0e217233
   0x0000000080000012 <entry+18>:       0e317233                .insn   4, 0x0e317233
(gdb) i r x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0xc                 0xc
(gdb) stepi       
0x000000008000000e      13          add x3, x0, x0     # Zero
   0x0000000080000002 <entry+2>:        000001b3                add     gp,zero,zero
   0x0000000080000006 <entry+6>:        0e215233                .insn   4, 0x0e215233
   0x000000008000000a <entry+10>:       0e315233                .insn   4, 0x0e315233
=> 0x000000008000000e <entry+14>:       0e217233                .insn   4, 0x0e217233
   0x0000000080000012 <entry+18>:       0e317233                .insn   4, 0x0e317233
(gdb) i r x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0x0                 0x0
(gdb) stepi       
0x0000000080000012      13          add x3, x0, x0     # Zero
   0x0000000080000002 <entry+2>:        000001b3                add     gp,zero,zero
   0x0000000080000006 <entry+6>:        0e215233                .insn   4, 0x0e215233
   0x000000008000000a <entry+10>:       0e315233                .insn   4, 0x0e315233
   0x000000008000000e <entry+14>:       0e217233                .insn   4, 0x0e217233
=> 0x0000000080000012 <entry+18>:       0e317233                .insn   4, 0x0e317233
(gdb) i r x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0x0                 0x0
(gdb) stepi       
18          nop
=> 0x0000000080000016 <entry+22>:       0001                    nop
(gdb) i r x2 x3 x4
x2             0xc                 0xc
x3             0x0                 0x0
x4             0xc                 0xc
```

It seems that everything worked as expected.

**Comment by rewired-gh** [comment]
The test for Zicond extension in https://github.com/riscv-non-isa/riscv-arch-test/pull/321 is some generated assembly code. And it looks like this,

```assembly
inst_607:
// rs1_val==-1431655765 and rs2_val==1431655766, 
// opcode: czero.eqz ; op1:x30; op2:x29; dest:x31; op1val:-0x55555555;  op2val:0x55555556
TEST_RR_OP(czero.eqz, x31, x30, x29, -0x55555555, -0x55555555, 0x55555556, x9, 71*XLEN/8, x10)
```

So we cannot directly substitute each line to make the test work. Since the latest RISC-V toolchain hasn't support Zicond extension, automated riscv-arch-test might not be possible, and putting CI in another drafted PR is not doable for now.

**Review by jerryz123 — APPROVED** [review]

File: src/main/scala/tile/BaseTile.scala [hunk 5]
```diff
@@ -117,6 +117,7 @@ trait HasNonDiplomaticTileParameters {
       Option.when(tileParams.core.useBitManip)(Seq("zbs")) ++
       Option.when(tileParams.core.useCryptoNIST)(Seq("zknd", "zkne", "zknh")) ++
       Option.when(tileParams.core.useCryptoSM)(Seq("zksed", "zksh")) ++
+      Option.when(tileParams.core.useConditionalZero)(Seq("zicond")) ++
```
Comment (ZenithalHourlyRate) [inline src/main/scala/tile/BaseTile.scala:120]: According to the isa spec, zicond should be put before zicsr and after zicntr.

>The first letter following the “Z” conventionally indicates the most closely related alphabetical
extension category, IMAFDQCVH. For the “Zam” extension for misaligned atomics, for exam-
ple, the letter “a” indicates the extension is related to the “A” standard extension. If multiple
"Z" extensions are named, they should be ordered first by category, then alphabetically within a
category—for example, “Zicsr Zifencei Zam”.

**Review by ZenithalHourlyRate — CHANGES_REQUESTED** [review]

**Review by ZenithalHourlyRate — APPROVED** [review]

**Comment by ZenithalHourlyRate** [comment]
@mergifyio backport master

**Pull request merged**